pkgs/by-name/im/immich/machine-learning.nix

```diff
@@ -0,0 +1,85 @@
+{
+  lib,
+  src,
+  immich,
+  python3,
+}:
+python3.pkgs.buildPythonApplication {
+  pname = "immich-machine-learning";
+  inherit (immich) version;
+  src = "${src}/machine-learning";
+  pyproject = true;
+
+  postPatch = ''
+    substituteInPlace pyproject.toml --replace-fail 'fastapi-slim' 'fastapi'
+
+    # Allow immich to use pydantic v2
+    substituteInPlace app/schemas.py --replace-fail 'pydantic' 'pydantic.v1'
+    substituteInPlace app/main.py --replace-fail 'pydantic' 'pydantic.v1'
+    substituteInPlace app/config.py \
+      --replace-fail 'pydantic' 'pydantic.v1'
+  '';
+
+  pythonRelaxDeps = [
+    "setuptools"
+    "pydantic"
+  ];
+  pythonRemoveDeps = [ "opencv-python-headless" ];
+
+  build-system = with python3.pkgs; [
+    poetry-core
+    cython
+  ];
+
+  dependencies =
+    with python3.pkgs;
+    [
+      insightface
+      opencv4
+      pillow
+      fastapi
+      uvicorn
+      aiocache
+      rich
+      ftfy
+      setuptools
+      python-multipart
+      orjson
+      gunicorn
+      huggingface-hub
+      tokenizers
+      pydantic
+    ]
+    ++ uvicorn.optional-dependencies.standard;
+
+  doCheck = false;
+
+  postInstall = ''
+    mkdir -p $out/share
+    cp log_conf.json $out/share
+
+    cp -r ann $out/${python3.sitePackages}/
+
+    makeWrapper ${lib.getExe python3.pkgs.gunicorn} $out/bin/machine-learning \
+      --prefix PYTHONPATH : "$out/${python3.sitePackages}:$PYTHONPATH" \
+      --set-default MACHINE_LEARNING_WORKERS 1 \
+      --set-default MACHINE_LEARNING_WORKER_TIMEOUT 120 \
+      --set-default MACHINE_LEARNING_CACHE_FOLDER /var/cache/immich \
+      --set-default IMMICH_HOST 127.0.0.1 \
+      --set-default IMMICH_PORT 3003 \
```

Comment:
Wouldn't it be better to set these defaults in `services.immich.machine-learning.environment`?

Reply:
If the package requires these to be set, it could still be a good idea to have them in the (or a) wrapper so the package could be ran without the NixOS module

Reply:
I don't think we have a precedent for wrapping packages with configuration envvars (other than ones we just made up ourselves, e.g. `NIXOS_OZONE_WL`). It'd be easy to cause subtle issues with the default upstream config and the user-set config by doing so, so I don't think we should start doing so, either.

Reply:
I think there are mulitple examples floating about in nixpkgs, but in any case this _is_ kind of options we've made up. They are used in the very next line:

```
-add-flags "app.main:app -k app.config.CustomUvicornWorker \
        -w \"\$MACHINE_LEARNING_WORKERS\" \
        -b \"\$IMMICH_HOST:\$IMMICH_PORT\" \
        -t \"\$MACHINE_LEARNING_WORKER_TIMEOUT\"
        --log-config-json $out/share/log_conf.json"
```

Having the wrapper is nice, since NixOS and nixpkgs should have separation. Giving this wrapper defaults then is a good courtesy.

If possible things in `pkgs` shouldn't be developed with too many assumptions around what the `nixos/` will do with them.

Reply:
> If possible things in `pkgs` shouldn't be developed with too many assumptions around what the `nixos/` will do with them.

Exactly, not-wrapping would provide the least assumptions.

Reply:
Having a wrapper is good, since it's running in a runtime.
Not having one would be like us shipping all java apps as plain jars.

unwrapped can also be made available, either through `<package>-unwrapped` or in a `passtrhu.unwrapped`

Reply:
> this _is_ kind of options we've made up

It's not: https://immich.app/docs/install/environment-variables#machine-learning

Reply:
Right, but it's not being consumed by the python script, this is recreating https://github.com/immich-app/immich/blob/30aa2c9b82a2817170cf275b24ad8b3021107bf1/machine-learning/start.sh#L17 inside our wrapper. The names chosen are just the same

Reply:
I'm also on team "do not set defaults in the wrapper". We should set those arguments in the module itself.

Reply:
Exposing a variant of the upstream script sounds reasonable. Not sure the NixOS service should depend on the nixpkgs wrapper's defaults though, it feels rather implicit. Should we just duplicate them in `serviceConfig`?